src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -12,99 +13,323 @@ namespace System.Net.Sockets
+                // Create the pipe for signalling shutdown, and register for "read" events for the pipe.  Now writing
+                // to the pipe will send an event to the event loop.
+                //
+                int* pipeFds = stackalloc int[2];
+                if (Interop.Sys.Pipe(pipeFds, Interop.Sys.PipeFlags.O_CLOEXEC) != 0)
+                {
+                    throw new InternalException();
+                }
+                _shutdownReadPipe = pipeFds[Interop.Sys.ReadEndOfPipe];
+                _shutdownWritePipe = pipeFds[Interop.Sys.WriteEndOfPipe];
 
-                for (int i = 0; i < numEvents; i++)
+                if (Interop.Sys.DangerousTryChangeSocketEventRegistration(_port, _shutdownReadPipe, Interop.Sys.SocketEvents.None, Interop.Sys.SocketEvents.Read, ShutdownHandle) != Interop.Error.SUCCESS)
                 {
-                    var handle = (GCHandle)(IntPtr)_buffer[i].Data;
-                    var context = (SocketAsyncContext)handle.Target;
+                    throw new InternalException();
+                }
+
+                //
+                // Start the event loop on its own thread.
+                //
+                Task.Factory.StartNew(
+                    EventLoop,
+                    CancellationToken.None,
+                    TaskCreationOptions.LongRunning,
+                    TaskScheduler.Default);
+            }
+            catch
+            {
+                FreeNativeResources();
+                throw;
+            }
+        }
+
+        private void EventLoop()
+        {
+            try
+            {
+                bool shutdown = false;
+                while (!shutdown)
+                {
+                    int numEvents = EventBufferCount;
+                    Interop.Error err = Interop.Sys.WaitForSocketEvents(_port, _buffer, &numEvents);
+                    if (err != Interop.Error.SUCCESS)
+                    {
+                        throw new InternalException();
+                    }
+
+                    // The native shim is responsible for ensuring this condition.
+                    Debug.Assert(numEvents > 0);
 
-                    if (context != null)
+                    for (int i = 0; i < numEvents; i++)
                     {
-                        context.HandleEvents(_buffer[i].Events);
+                        IntPtr handle = _buffer[i].Data;
+                        if (handle == ShutdownHandle)
+                        {
+                            shutdown = true;
+                        }
+                        else
+                        {
+                            SocketAsyncContext context = GetContextFromHandle(handle);
+                            if (context != null)
+                            {
+                                context.HandleEvents(_buffer[i].Events);
+                            }
+                        }
                     }
                 }
+
+                FreeNativeResources();
```

Comment:
Nit: I realize we're going to fail fast if an exception occurs, but you might still want to put this in the finally.  I'm just imagining some time in the future we decide to change the FailFast to something less drastic, and at that point there'll be a leak if this isn't also moved to a finally.  Up to you, though.